pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go

```diff
@@ -430,31 +430,67 @@ func patchHTTPFilters(patchContext networking.EnvoyFilter_PatchContext,
 		}
 		patchHTTPFilter(patchContext, filterKey, patches, listener, fc, filter, httpFilter, &httpFiltersRemoved)
 	}
+	opatches := make([]*model.EnvoyFilterConfigPatchWrapper, 0) // List of final patches to be processed.
+	mpatches := make([]*model.EnvoyFilterConfigPatchWrapper, 0) // List of matched patches based on patch context.
 	applied := false
-	for _, cp := range patches[networking.EnvoyFilter_HTTP_FILTER] {
-		if !commonConditionMatch(patchContext, cp) ||
-			!listenerMatch(listener, cp) ||
-			!filterChainMatch(listener, fc, cp) ||
-			!networkFilterMatch(filter, cp) {
+
+	// First collect all matched patches.
+	for _, lp := range patches[networking.EnvoyFilter_HTTP_FILTER] {
+		if !commonConditionMatch(patchContext, lp) ||
+			!listenerMatch(listener, lp) ||
+			!filterChainMatch(listener, fc, lp) ||
+			!networkFilterMatch(filter, lp) {
 			continue
 		}
-		if cp.Operation == networking.EnvoyFilter_Patch_ADD {
+		mpatches = append(mpatches, lp)
+	}
+
+	// Identify cyclic patches and isolate them.
+	iapatches := make([]*model.EnvoyFilterConfigPatchWrapper, 0)
+	for _, lp := range mpatches {
+		if !(lp.Operation == networking.EnvoyFilter_Patch_INSERT_BEFORE || lp.Operation == networking.EnvoyFilter_Patch_INSERT_AFTER) {
+			continue
+		}
+		if hasHTTPFilterMatch(lp) {
+			iapatches = append(iapatches, lp)
+		}
+	}
+	if len(iapatches) > 0 {
+		// Detect cycles and order them if we have insert after and insert before patches.
+		cycles := hasCyclicDependencies(iapatches)
+		if cycles {
+			IncrementEnvoyFilterMetric(filterKey, HttpFilter, applied)
+			log.Warnf("envoy filter %s has patches that have cyclic dependencies. So filter is skipped", filterKey)
```

Comment:
What is the filterkey here? Is it enough to identify the problem ?

Reply:
we still want to apply independent patches?
Adding a cycle should not remove all filters.